docs/resources/rules/assertions.md

```diff
@@ -0,0 +1,393 @@
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.x-codeSamples
+        message: x-codeSamples must be defined
+        severity: error
+        defined: true
+```
+
+### `undefined` example
+
+The following example asserts that `x-code-samples` is undefined.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.x-code-samples
+        message: x-code-samples is deprecated
+        suggest: 
+          - x-codeSamples instead of x-code-samples
+        severity: error
+        undefined: true
+```
+
+### `nonEmpty` example
+
+The following example asserts that the operation summary is not empty.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.summary
+        message: Operation summary should not be empty
+        severity: error
+        nonEmpty: true
+```
+
+### `minLength` example
+
+The following example asserts that the minimum length of each operation summary is 20 characters.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.summary
+        message: Operation summary must have minimum of 20 chars length
+        severity: error
+        minLength: 20
+```
+
+### `maxLength` example
+
+The following example asserts that the maximum length of each operation summary is 20 characters.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.summary
+        message: Operation summary must have maximum of 20 characters
+        severity: error
+        maxLength: 20
+```
+
+### `sortOrder` example
+
+Sort order can be applied on lists (arrays) of strings or lists of objects (collections).
+
+The following example asserts that the status schema's enum list is in alphabetical order.
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Schema.status.enum
+        message: The status enums should be in alphabetical order
+        suggest:
+          - alphabetize the list of status enums
+        severity: error
+        sortOrder: asc
+```
+
+The following example asserts the tag names (tags are a collection) are sorted in alphabetical order.
+
+The `property` field is **REQUIRED** to sort collections based on that property's value. 
+
+```yaml
+lint:
+  rules:
+    assertions:
+      - on: Operation.tags
+        message: Tags should be ordered by name in alphabetical order
+        severity: error
+        sortOrder:
+          direction: asc
+          property: name
+```
+
+## OpenAPI node types
+
+Redocly defines a type tree based on the document type.
+OpenAPI 2.0 has one type tree.
+OpenAPI 3.0 and OpenAPI 3.1 share a type tree.
+
+Node types are traversed using dot notation.
+
+The following example shows ways to traverse to the equivalent destination.
+One starts from the definition root into the info description.
+The other starts from the info section.
+
+```yaml Traversal from DefinitionRoot object type
+lint:
+  rules:
+    assertions:
+      - on:
+        - DefinitionRoot.info.description
```

Comment:
```suggestion
        - DefinitionRoot.Info.description
```

Reply:
I don't understand. Can you traverse objects but not regular properties?

Reply:
No, this is not about traversal at all. I'll explain you on a call.